### PR TITLE
Replace useSafeGasEstimatePolling with usePolling

### PR DESCRIPTION
**Pull request by jiexi**
## **Description**
Added `usePolling` hook and used it within `useSafeGasEstimatePolling` hook.

This replaces where gas estimates are coming from for the confirmation flow.

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/2014

## **Manual testing steps**

1. Hit send button
2. get to confirmation screen
3. see that gas estimation updates

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/hooks/useGasFeeEstimates.js [hunk 1]
```diff
@@ -27,12 +29,18 @@ import { useSafeGasEstimatePolling } from './useSafeGasEstimatePolling';
  *
  * @returns {GasEstimates} GasEstimates object
  */
-export function useGasFeeEstimates() {
+export function useGasFeeEstimates(networkClientId) {
   const gasEstimateType = useSelector(getGasEstimateType);
   const gasFeeEstimates = useSelector(getGasFeeEstimates, isEqual);
```
Comment (shanejonas) [inline ui/hooks/useGasFeeEstimates.js:34]: should this be using `networkClientId` or `chainId` or something in the selector

File: ui/ducks/metamask/metamask.js [hunk 2]
```diff
@@ -331,12 +331,14 @@ export function isNotEIP1559Network(state) {
  * Function returns true if network details are fetched and it is found to support EIP-1559
  *
  * @param state
+ * @param networkClientId
```
Comment (shanejonas) [inline ui/ducks/metamask/metamask.js:334]: should add jsdoc for optional param

**Review by shanejonas — DISMISSED** [review]

File: ui/pages/confirmations/components/advanced-gas-fee-popover/advanced-gas-fee-inputs/base-fee-input/base-fee-input.js [hunk 3]
```diff
@@ -59,23 +59,24 @@ const BaseFeeInput = () => {
...
+    estimateUsed !== PriorityLevels.custom &&
+    advancedGasFeeValues?.maxBaseFee &&
+    editGasMode !== EditGasModes.swaps
+      ? advancedGasFeeValues.maxBaseFee
+      : maxFeePerGas;
+
+  const [baseFee, setBaseFee] = useState(defaultBaseFee);
+  useEffect(() => {
+    setBaseFee(defaultBaseFee);
+  }, [defaultBaseFee, setBaseFee]);
```
Comment (jiexi) [inline ui/pages/confirmations/components/advanced-gas-fee-popover/advanced-gas-fee-inputs/base-fee-input/base-fee-input.js:79]: Pointing out this change in particular. Previously there were assumptions that the values that make up the default baseFee value would be set correctly on first render. This is no longer true. Because of that, we need to use an effect to update baseFee after the fact. A potential problem is that this also introduces a new unintended behavior that baseFee will continue to be updated each time any of the underlying values that are used to determine it change.

File: ui/hooks/useGasFeeEstimates.js [hunk 4]
```diff
@@ -25,14 +32,44 @@ import { useSafeGasEstimatePolling } from './useSafeGasEstimatePolling';
...
+    getIsGasEstimatesLoadingByChainId(state, { chainId, networkClientId }),
+  );
+  const isNetworkBusy = useSelector((state) =>
+    getIsNetworkBusyByChainId(state, chainId),
+  );
+
+  useEffect(() => {
+    getNetworkConfigurationByNetworkClientId(networkClientId).then(
+      (networkConfig) => {
+        if (networkConfig) {
```
Comment (adonesky1) [inline ui/hooks/useGasFeeEstimates.js:61]: could this be undefined?

File: ui/hooks/useGasFeeEstimates.js [hunk 5]
```diff
@@ -25,14 +32,44 @@ import { useSafeGasEstimatePolling } from './useSafeGasEstimatePolling';
...
  */
-export function useGasFeeEstimates() {
-  const gasEstimateType = useSelector(getGasEstimateType);
-  const gasFeeEstimates = useSelector(getGasFeeEstimates, isEqual);
-  const isGasEstimatesLoading = useSelector(getIsGasEstimatesLoading);
-  const isNetworkBusy = useSelector(getIsNetworkBusy);
-  useSafeGasEstimatePolling();
+export function useGasFeeEstimates(_networkClientId) {
+  const selectedNetworkClientId = useSelector(getSelectedNetworkClientId);
+  const networkClientId = _networkClientId ?? selectedNetworkClientId;
```
Comment (adonesky1) [inline ui/hooks/useGasFeeEstimates.js:40]: should we throw an error if somehow this isn't defined?

File: ui/hooks/useGasFeeEstimates.js (hunk 4 above)
Comment (adonesky1) [reply ui/hooks/useGasFeeEstimates.js:61]: doesn't seem like we have any plan in the case that its not... should we throw an error?

File: ui/pages/confirmations/components/advanced-gas-fee-popover/advanced-gas-fee-inputs/base-fee-input/base-fee-input.js (hunk 3 above)
Comment (adonesky1) [reply ui/pages/confirmations/components/advanced-gas-fee-popover/advanced-gas-fee-inputs/base-fee-input/base-fee-input.js:79]: > A potential problem is that this also introduces a new unintended behavior that baseFee will continue to be updated each time any of the underlying values that are used to determine it change.

Is this a problem? The baseFee we show should be updating as these values change, no?

Comment (adonesky1) [reply ui/pages/confirmations/components/advanced-gas-fee-popover/advanced-gas-fee-inputs/base-fee-input/base-fee-input.js:79]: Tested this and seems to work well!

File: ui/hooks/useGasFeeEstimates.js (hunk 5 above)
Comment (jiexi) [reply ui/hooks/useGasFeeEstimates.js:40]: i think we just let it blow up downstream. selectedNetworkClientId coming back undefined would signal something worse is messed up

File: .devcontainer/devcontainer.json [hunk 6]
```diff
@@ -0,0 +1,58 @@
+{
```
Comment (shanejonas) [inline .devcontainer/devcontainer.json:1]: this file/folder was committed by mistake i think

Comment (shanejonas) [reply .devcontainer/devcontainer.json:1]: ah i see, we need to rebase my original branch

File: test/data/mock-send-state.json [hunk 7]
```diff
@@ -24,6 +25,9 @@
         "name": null
       }
     },
+    "showIpfsModalOpen": false,
+    "showKeyringRemovalSnapModal": false,
+    "showWhatsNewPopup": false,
```
Comment (adonesky1) [inline test/data/mock-send-state.json:30]: just curious, seems like a there are few things added here that are unrelated to the gasFee polling, is there a reason they were needed as part of this PR?

File: test/e2e/snaps/test-snap-txinsights-v2.spec.js [hunk 8]
```diff
@@ -99,6 +99,12 @@ describe('Test Snap TxInsights-v2', function () {
           WINDOW_TITLES.Dialog,
           windowHandles,
         );
+
+        await driver.findClickableElement({
+          text: 'Confirm',
+          tag: 'button',
+        });
```
Comment (adonesky1) [inline test/e2e/snaps/test-snap-txinsights-v2.spec.js:106]: could you help me understand why this change was required?

File: test/data/mock-send-state.json (hunk 7 above)
Comment (jiexi) [reply test/data/mock-send-state.json:30]: wrapping renders in `act` catches previous uncaught state selection failures

File: test/e2e/snaps/test-snap-txinsights-v2.spec.js (hunk 8 above)
Comment (jiexi) [reply test/e2e/snaps/test-snap-txinsights-v2.spec.js:106]: It's all rooted in the issue of not having correct gas estimate values on first render from the hook coupled with gas estimation logic being mixed into the ui components themselves. If the gas estimation UI (the details tab) isn't mounted when the gas estimates become available, then the gas estimates don't get updated. I can't think of a good way to deal with this issue as gas estimation shouldn't be in the ui layer in the first place. @matthewwalsh0 has mentioned tackling this by moving gas estimation into the background at the TransactionController layer

File: ui/hooks/useGasFeeEstimates.js (hunk 4 above)
Comment (adonesky1) [reply ui/hooks/useGasFeeEstimates.js:64]: If its `undefined` we could set some error state that we could return from the hook. The reason I think we probably need to do something like this is because it won't be apparent what's wrong and why polling isn't working if this failed... Though I admit I don't see why it would fail

File: test/e2e/snaps/test-snap-txinsights-v2.spec.js (hunk 8 above)
Comment (adonesky1) [reply test/e2e/snaps/test-snap-txinsights-v2.spec.js:106]: 🤔 ok this is a much bigger topic than the initial question here but I think ultimately we **do** want the polling to be tied to UI components. See this [thread](https://consensys.slack.com/archives/C06KYQG05A4/p1708980778877819)

**Comment by tmashuang** [comment]
Overall no issues with behavior from my first pass. Although there are styling issues that came up on a certain network.

Network name: Ganache 8545 1337
New RPC URL: HTTP://127.0.0.1:8545
Chain ID: 1337
Currency symbol: g-eth

ganache-send-tx-gas-styling-issue

File: test/e2e/snaps/test-snap-txinsights-v2.spec.js (hunk 8 above)
Comment (adonesky1) [reply test/e2e/snaps/test-snap-txinsights-v2.spec.js:106]: > Tied to meaning started and stopped based on UI component mounting, yes I agree. Tied to meaning the actual gas estimates costs for a transactions being calculated by tightly coupled react hooks, no I disagree.

🤝 we are in accord.

**Comment by seaona** [comment]
I will update this same comment with the findings during the testing phase.

## QA Findings
### 1. Unresponsive network handling behaviour
Whenever I am on a network which is unresponsive, I see errors on
- `StaticIntervalPollingController`
- `PollingBlockTracker`
- RPC errors

Could those be handled somehow? This way we would avoid having too much noise in Sentry.
Then, after I switch to another responsive network, I continue to see errors and requests being made in the old network. 


https://github.com/MetaMask/metamask-extension/assets/54408225/aab0c11e-abc8-4920-861e-29600013a964




https://github.com/MetaMask/metamask-extension/assets/54408225/c520e259-ef0d-4a93-8fa1-f08885e47666


## 2. Couldn't find networkClientId for chainId
I have encountered the background error: `Couldn't find networkClientId for chainId` but still not sure yet how to repro. I was using a local network.

Screenshot from 2024-03-03 12-48-44

**Comment by jiexi** [comment]
`Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application` should be addressed by https://github.com/MetaMask/metamask-extension/pull/23010/commits/d0d702f3f21b4be3ea6fa596205495eca65d1f05

**Comment by jiexi** [comment]
Item 3: Seems like this is caused by having a very large state in MetaMask, likely from having many transactions in state and/or a slow machine. IMO not related to these particular set of changes.

Item 4: This is a DetectTokensController issue that I've brought to the attention of the appropriate individuals.


That should be all the items. Thank you for this thoroughly detailed list, @seaona!

**Comment by seaona** [comment]
### 1. :green_circle:  Unresponsive network handling behaviour --> out of scope

> I wasn't able to reproduce this on this branch 🤔 Could you describe any other steps you took before you saw this happening? I'm assuming you just started a local testnet and then turned it off and then looked for errors and then switched networks to sepolia?

I am able to repro, I see errors both from `StaticIntervalPollingController` and `PolliognBlockTracker` after I switch networks. However, as you pointed then those seem to be independent from this PR, so it can be left out of scope if that sounds good. I marked is as green, but feel free to disagree

**Steps:**
1. Add ganache (server is on)
2. Perform a tx
3. Switch to sepolia
4. See errors continue --only stopped if I reload the wallet with the reload button from the Extension Chrome page

### 2. Can't perform a React state update on an unmounted component
- :green_circle: I see the `useGasFeeEstimates` fixed
- :yellow_circle: I still see the `gas-timing-component` when editing a transaction



https://github.com/MetaMask/metamask-extension/assets/54408225/30035069-b128-46e4-9d6a-1ec8baa935f8

### 3. Performance Degradation
It seems that this might be also out of scope of this PR. I've seen that this happened to other branches with multichain flag enabled. Another QA engineer from Confirmations (@sleepytanya ) also experienced this in other branches.

### 4. Couldn't find networkClientId for chainId
This is out of scope as per this comment:
> Item 4: This is a DetectTokensController issue that I've brought to the attention of the appropriate individuals.

**Comment by jiexi** [comment]
2. Can't perform a React state update on an unmounted component
  - I still see the gas-timing-component when editing a transaction

This is a bug that currently exists in develop. I have manually checked on develop to confirm. We will address as part of this PR though https://github.com/MetaMask/metamask-extension/pull/23010/commits/1080c91aa5f0077229bf270b5b2916b61899d305

**Comment by adonesky1** [comment]
> I am able to repro, I see errors both from StaticIntervalPollingController and PolliognBlockTracker after I switch networks. 

> Steps:
> 
> Add ganache (server is on)
> Perform a tx
> Switch to sepolia

I agree this is should be addressed separately but I'm confused why I'm not seeing the errors you describe. I followed the steps above and this is what I'm seeing:

https://github.com/MetaMask/metamask-extension/assets/34557516/546a8592-3237-44d1-af49-27d02e8d6a20

File: ui/pages/confirmations/components/gas-timing/gas-timing.component.js [hunk 9]
```diff
@@ -65,6 +65,7 @@ export default function GasTiming({
   const previousIsUnknownLow = usePrevious(isUnknownLow);
```
Comment (adonesky1) [inline ui/pages/confirmations/components/gas-timing/gas-timing.component.js:65]: Looks like we're not reading `gasFeeEstimates` and `gasEstimateType` from the `gasFeeEstimatesByChainId` state yet on this component. Is that intentional?

Comment (jiexi) [reply ui/pages/confirmations/components/gas-timing/gas-timing.component.js:65]: i believe values derived from gasFeeEstimatesByChainId are used here

**Review by adonesky1 — DISMISSED** [review]

File: ui/hooks/usePolling.test.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    const options = {};
+    const mockState = {
+      metamask: {},
+    };
+
+    const wrapper = ({ children }) => (
+      <>
+        <Provider store={configureStore(mockState)}>{children}</Provider>
+      </>
+    );
```
Comment (jpuri) [inline ui/hooks/usePolling.test.js:24]: We have test utility function `renderHookWithProvider` already defined. May be can be used here.

File: ui/hooks/usePolling.ts [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+      isMounted = false;
+      cleanup();
+    };
+  }, [
+    usePollingOptions.networkClientId,
+    usePollingOptions.options &&
+      JSON.stringify(
+        usePollingOptions.options,
+        Object.keys(usePollingOptions.options).sort(),
+      ),
```
Comment (jpuri) [inline ui/hooks/usePolling.ts:52]: Wondering if its possible to simplify above conditional.

File: ui/pages/confirmations/components/advanced-gas-fee-popover/advanced-gas-fee-inputs/priority-fee-input/priority-fee-input.js [hunk 12]
```diff
@@ -66,19 +66,20 @@ const PriorityFeeInput = () => {
...
+    estimateUsed !== PriorityLevels.custom &&
+    advancedGasFeeValues?.priorityFee &&
+    editGasMode !== EditGasModes.swaps
+      ? advancedGasFeeValues.priorityFee
+      : maxPriorityFeePerGas;
+
+  const [priorityFee, setPriorityFee] = useState(defaultPriorityFee);
+  useEffect(() => {
+    setPriorityFee(defaultPriorityFee);
+  }, [defaultPriorityFee, setPriorityFee]);
```
Comment (jpuri) [inline ui/pages/confirmations/components/advanced-gas-fee-popover/advanced-gas-fee-inputs/priority-fee-input/priority-fee-input.js:82]: Here we did calculation in `useState` as this should be done only once. `defaultPriorityFee` should not change.

File: ui/hooks/usePolling.test.js (hunk 10 above)
Comment (adonesky1) [reply ui/hooks/usePolling.test.js:24]: Looking into this!

Comment (adonesky1) [reply ui/hooks/usePolling.test.js]: Done here: https://github.com/MetaMask/metamask-extension/pull/23010/commits/02a0c73d28e7b61dbaeb91589d5bb9a9b98aa129

File: ui/hooks/usePolling.ts (hunk 11 above)
Comment (adonesky1) [reply ui/hooks/usePolling.ts:52]: This logic is to ensure that if the `usePollingOptions.options` is present we can guarantee the order of the keys in the option object so that the hook doesn't rerun unnecessarily. I'm not sure if there is a better way to express that conditional.

File: ui/pages/confirmations/components/advanced-gas-fee-popover/advanced-gas-fee-inputs/priority-fee-input/priority-fee-input.js (hunk 12 above)
Comment (adonesky1) [reply ui/pages/confirmations/components/advanced-gas-fee-popover/advanced-gas-fee-inputs/priority-fee-input/priority-fee-input.js:82]: I believe this is necessary because with the new polling hook some of the values from `useGasFeeContext` used to derive `defaultPriorityFee` will be undefined on first render. @jiexi can maybe add some color here

Comment (jpuri) [reply ui/pages/confirmations/components/advanced-gas-fee-popover/advanced-gas-fee-inputs/priority-fee-input/priority-fee-input.js:82]: But we need to check scenario if user has set a custom value for priority fee, can this override that ?

Comment (jiexi) [reply ui/pages/confirmations/components/advanced-gas-fee-popover/advanced-gas-fee-inputs/priority-fee-input/priority-fee-input.js:82]: User set Max Base Fee and Priority Fee are not reset/changed by this. I have verified by delaying the gas estimate state update in the hook and seeing that the Max Fee does not change and that of course the Max Base Fee and Priority Fee inputs do not change either

**Review by adonesky1 — APPROVED** [review]

**Review by jpuri — APPROVED** [review]

**Review by Mrtenz — APPROVED** [review]
